src/main/java/org/codehaus/plexus/util/io/CachingOutputStream.java

```diff
@@ -0,0 +1,66 @@
+package org.codehaus.plexus.util.io;
+
+/*
+ * Copyright The Codehaus Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Caching OutputStream to avoid overwriting a file with
+ * the same content.
+ */
+public class CachingOutputStream extends ByteArrayOutputStream
+{
+    private final Path path;
+    private boolean modified;
+
+    public CachingOutputStream( File path )
+    {
+        this( Objects.requireNonNull( path ).toPath() );
+    }
+
+    public CachingOutputStream( Path path )
+    {
+        this.path = Objects.requireNonNull( path );
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        byte[] data = toByteArray();
+        if ( Files.exists( path ) && Files.size( path ) == data.length )
+        {
+            byte[] old = Files.readAllBytes( path );
+            if ( Arrays.equals( old, data ) )
```

Comment:
this could be a lot of data up in memory in case of huge files?

Reply:
Yes, that's right.  That's one of the reason why I originally did not want to put those classes in a utils library, so that it does not get reused in bad situations.
The first fix would be to add a bit of javadoc to make the possible problems more visible.
A possible enhancement to work around the problem would be to use a threshold (big enough by default) above which the content would just be written to the file without reading it first.  However, this would still force the new content to be buffered in memory.
Another one would be to use a mechanism similar to the one that has been implemented for resource filtering, see https://github.com/apache/maven-shared-utils/blob/master/src/main/java/org/apache/maven/shared/utils/io/FileUtils.java#L1946-L1991, i.e. read and write the file at the same time and switch to direct write once a change has been found.  This would have the benefit of supporting large files, without the need to buffer more than a few kilobytes (just for the sake of optimising the IO), with the drawbacks of introducing a much more complex code.

Reply:
I'll give a try to the last solution which would obviously be the best.

Reply:
thanks!